crates/networking/p2p/sync.rs

```diff
@@ -114,48 +128,57 @@ impl SyncManager {
-                // - Fetch each block's state via snap p2p requests
-                // - Fetch each blocks and its receipts via eth p2p requests
-                // TODO: We are currently testing against our implementation that doesn't hold an independant snapshot and can provide all historic state
-                //       We should fetch all available state and then resort to state healing to fetch the rest
-                let (bytecode_sender, bytecode_receiver) = mpsc::channel::<Vec<H256>>(500);
-                let mut set = tokio::task::JoinSet::new();
-                set.spawn(bytecode_fetcher(
-                    bytecode_receiver,
-                    self.peers.clone(),
-                    store.clone(),
-                ));
-                set.spawn(fetch_blocks_and_receipts(
+                // - Fetch each block's body and its receipt via eth p2p requests
+                // - Fetch the pivot block's state via snap p2p requests
+                // - Execute blocks after the pivote (like in full-sync)
+                let fetch_bodies_handle = tokio::spawn(fetch_block_bodies(
                     all_block_hashes.clone(),
                     self.peers.clone(),
-                    store.clone(),
-                ));
-                let state_roots = all_block_headers
-                    .iter()
-                    .map(|header| header.state_root)
-                    .collect::<Vec<_>>();
-                set.spawn(fetch_snap_state(
-                    bytecode_sender,
-                    state_roots.clone(),
-                    self.peers.clone(),
-                    store.clone(),
                 ));
-                // Store headers
-                let mut latest_block_number = 0;
-                for (header, hash) in all_block_headers
-                    .into_iter()
-                    .zip(all_block_hashes.into_iter())
-                {
-                    latest_block_number = header.number;
-                    store.set_canonical_block(header.number, hash)?;
-                    store.add_block_header(hash, header)?;
+                let mut pivot_idx = if all_block_headers.len() > MIN_FULL_BLOCKS {
+                    all_block_headers.len() - MIN_FULL_BLOCKS
+                } else {
+                    all_block_headers.len() - 1
+                };
+                let mut pivot_root = all_block_headers[pivot_idx].state_root;
+                let mut pivot_number = all_block_headers[pivot_idx].number;
+
+                let mut stale_pivot =
+                    !rebuild_state_trie(pivot_root, self.peers.clone(), store.clone()).await?;
+                // If the pivot became stale, set a further pivot and try again
+                if stale_pivot && pivot_idx != all_block_headers.len() - 1 {
+                    warn!("Stale pivot, switching to newer head");
+                    pivot_idx = all_block_headers.len() - 1;
```

Comment:
shouldn't you update `all_block_headers` before choosing a new pivot? I imagine the state_pivot could happen after some time and there are a lot more headers.

Reply:
why are you choosing `pivot_idx = all_block_headers.len() - 1;`? It will only move forward by 63 blocks.

Reply:
We are not fetching any more headers as we don't know about any later block aside from the sync root, the only way to sync further would be to wait for a next fork choice update.